libraries/SITL/SIM_AirSim.cpp

```diff
@@ -0,0 +1,158 @@
+/* 
+	Simulator Connector for AirSim
+*/
+
+#include "SIM_AirSim.h"
+
+#include <stdio.h>
+#include <arpa/inet.h>
+#include <errno.h>
+
+#include <AP_HAL/AP_HAL.h>
+
+extern const AP_HAL::HAL& hal;
+
+using namespace SITL;
+
+AirSim::AirSim(const char *home_str, const char *frame_str) :
+	Aircraft(home_str, frame_str),
+	sensor_sock(true),
+	control_sock(true),
+	last_timestamp(0),
+	opened_control_socket(false),
+	opened_sensor_socket(false)
+{
+	printf("Starting Airsim at %s:%u:%u\n", airsim_ip, airsim_sensor_port, airsim_control_port);
+}
+
+/*
+	Connect to required sockets
+*/
+bool AirSim::connect_sensor_socket(void)
+{
+	if (opened_sensor_socket) {
+		return true;
+	}
+	// This should be binding to 127.0.0.1 or 0.0.0.0 ?
+	if(!sensor_sock.bind(airsim_ip, airsim_sensor_port)) {
+		printf("Unable to open Airsim sensor socket at %s:%u\n", airsim_ip, airsim_sensor_port);
+		printf("Error: %s\n", strerror(errno));
+		return false;
+	}
+	printf("Opened Airsim sensor socket at %s:%u\n", airsim_ip, airsim_sensor_port);
+	sensor_sock.set_blocking(false);
+	sensor_sock.reuseaddress();
+	opened_sensor_socket = true;
+
+	return true;
+}
+
+bool AirSim::connect_control_socket(void)
+{
+	if (opened_control_socket) {
+		return true;
+	}
+	// binding to 0.0.0.0 or 127.0.0.1 ?
+	if(!control_sock.bind(airsim_ip, airsim_control_port)) {
+		printf("Unable to open Airsim control socket at %s:%u\n", airsim_ip, airsim_control_port);
+		printf("Error: %s\n", strerror(errno));
+		return false;
+	}
+	printf("Opened Airsim control socket at %s:%u\n", airsim_ip, airsim_control_port);
+	control_sock.set_blocking(false);
+	control_sock.reuseaddress();
+	opened_control_socket = true;
+
+	return true;
+}
+
+void AirSim::send_servos(const struct sitl_input &input)
+{
+	servo_packet pkt;
+
+	for(int i=0; i<kArduCopterRotorControlCount; ++i) {
+		// No need to normalize since Airsim does that in ArduCopterSoloApi
+		// Range 1000-2000
+		pkt.pwm[i] = input.servos[i];
+	}
+	pkt.speed = input.wind.speed;
+	pkt.direction = input.wind.direction;
+	pkt.turbulance = input.wind.turbulence;
+
+	control_sock.send(&pkt, sizeof(pkt));
+}
+
+/*
+	Airsim sends in little endian
+*/
+void fdm_packet::ByteSwap(void)
+{
+	uint64_t *buf = (uint64_t *)this;
+	for (uint16_t i=0; i<sizeof(*this)/8; ++i) {
+		buf[i] = le64toh(buf[i]);
+	}
+}
+
+/*
+	Recieve new sensor data from simulator
+	This is a blocking function
+*/
+void AirSim::recv_fdm(const struct sitl_input &input)
+{
+	fdm_packet pkt;
+
+	// Resend servo packet every 0.1 second till we get data
+	// We do get the timestamp in the packet, sync can be achieved in a similar manner as JSBSim
+	while (sensor_sock.recv(&pkt, sizeof(pkt), 100) != sizeof(pkt)) {
+		send_servos(input);
+	}
+	// NOTE: Airsim sends data in little-endian
+
+	// Taken from CRRCSim.cpp
```

Comment:
Does AirSim conform to NED?

Reply:
Need to confirm that. I didn't fix all this till now since I was focusing on making sure that they are able to communicate first and then I could proceed ahead

Reply:
https://github.com/microsoft/AirSim/blob/master/docs/apis.md#coordinate-system
Internally, no. Fixed that while calculating position